Core/include/Acts/Propagator/Navigator.hpp

```diff
@@ -248,937 +259,483 @@ class Navigator {
+    if (state.currentVolume == nullptr) {
       ACTS_VERBOSE(volInfo(state) << "No volume found, stop navigation.");
-      // Set navigation break and release the navigation step size
-      state.navigation.navigationBreak = true;
-      stepper.releaseStepSize(state.stepping, ConstrainedStep::actor);
-      return;
+      state.navigationBreak = true;
+      return NavigationTarget::invalid();
     }
 
-    state.navigation.currentLayer =
-        state.navigation.currentVolume->associatedLayer(
-            state.geoContext, stepper.position(state.stepping));
+    state.currentLayer = state.currentVolume->associatedLayer(
+        state.options.geoContext, position);
 
     ACTS_VERBOSE(volInfo(state) << "Resolved volume and layer.");
 
     // Rerun the targeting
-    if (tryTargetNextSurface()) {
-      return;
+    nextTarget = tryEstimateNextTarget();
+    if (nextTarget.isValid()) {
+      return nextTarget;
     }
 
-    ACTS_VERBOSE(volInfo(state) << "No targets found again, we got "
-                                   "really lost! Stop navigation.");
-    // Set navigation break and release the navigation step size
-    state.navigation.navigationBreak = true;
-    stepper.releaseStepSize(state.stepping, ConstrainedStep::actor);
+    ACTS_VERBOSE(
+        volInfo(state)
+        << "No targets found again, we got really lost! Stop navigation.");
+    state.navigationBreak = true;
+    return NavigationTarget::invalid();
   }
 
-  /// @brief Navigator post step call
-  ///
-  /// (a) It initializes the Navigation stream if start volume is
-  ///     not yet defined:
-  ///  - initialize the volume
-  ///  - establish the start layer and start volume
-  ///  - set the current surface to the start surface
+  /// @brief Check if the current target is still valid
   ///
-  /// (b) It establishes the currentSurface status during
-  ///     the propagation flow, currentSurface can be
-  ///  - surfaces still to be handled within a layer
-  ///  - layers still to be handled within a volume
-  ///  - boundaries still to be handled to exit a volume
+  /// This function checks if the target is valid.
   ///
-  /// @tparam propagator_state_t is the type of Propagatgor state
-  /// @tparam stepper_t is the used type of the Stepper by the Propagator
+  /// @param state The navigation state
+  /// @param position The current position
+  /// @param direction The current direction
   ///
-  /// @param [in,out] state is the mutable propagator state object
-  /// @param [in] stepper Stepper in use
-  template <typename propagator_state_t, typename stepper_t>
-  void postStep(propagator_state_t& state, const stepper_t& stepper) const {
-    // Check if the navigator is inactive
-    if (inactive(state, stepper)) {
-      return;
-    }
-
-    // Set the navigation stage
-    state.navigation.navigationStage = Stage::undefined;
+  /// @return True if the target is valid
+  bool checkTargetValid(const State& state, const Vector3& position,
```

Comment:
naming: `haveValidTarget`?

Reply:
This is a callback to the navigator to check if the current target is still valid if it was not reached by the first step. For potential future implementations this can be used invalidate the state and to cause a reinitialization.